cmd/uzfs_test/zrepl_utest.c

```diff
@@ -0,0 +1,411 @@
+{
+
+	int i;
+
+	for (i = 0; i < size; i++) {
+		if (p[i] != 'C') {
+			return (-1);
+		}
+	}
+	return (0);
+}
+
+static void
+reader_thread(void *arg)
+{
+
+	char *buf;
+	int sfd, count;
+	kmutex_t *mtx;
+	kcondvar_t *cv;
+	int *threads_done;
+	int write_ack_cnt = 0;
+	int read_ack_cnt = 0;
+	zvol_io_hdr_t *hdr;
+	worker_args_t *warg = (worker_args_t *)arg;
+
+	mtx = warg->mtx;
+	cv = warg->cv;
+	threads_done = warg->threads_done;
+
+	sfd = warg->sfd;
+	hdr = kmem_alloc(sizeof (zvol_io_hdr_t), KM_SLEEP);
+	buf = kmem_alloc(warg->io_block_size, KM_SLEEP);
+
+	printf("Start reading ........\n");
+	while (1) {
+		if ((warg->max_iops == write_ack_cnt) &&
+		    (warg->max_iops == read_ack_cnt)) {
+			break;
+		}
+		count = read(sfd, (void *)hdr, sizeof (zvol_io_hdr_t));
+		if (count == -1) {
+			printf("Read error reader_thread\n");
+			break;
+		}
+
+		if (hdr->opcode == ZVOL_OPCODE_WRITE) {
+			write_ack_cnt++;
+			bzero(hdr, sizeof (zvol_io_hdr_t));
+			continue;
+		}
+
+		if (hdr->opcode == ZVOL_OPCODE_READ) {
+			read_ack_cnt++;
+
+			int nbytes = hdr->len;
+			char *p = buf;
+			while (nbytes) {
+				count = read(sfd, (void *)p, nbytes);
+				if (count == -1) {
+					printf("\n");
+					printf("Read error in reader_thread "
+					    "reading data\n");
+				}
+				p += count;
+				nbytes -= count;
+			}
+
+			if (zrepl_verify_data(buf, warg->io_block_size) == -1) {
+				printf("Read :%d bytes data\n", count);
+				printf("Data mismatch\n");
+			}
+		}
+
+		bzero(hdr, sizeof (zvol_io_hdr_t));
+		bzero(buf, warg->io_block_size);
+	}
+
+	printf("Total iops requested:%d writes:%d total reads: %d\n",
+	    warg->max_iops, write_ack_cnt, read_ack_cnt);
+	free(hdr);
+	free(buf);
+	mutex_enter(mtx);
+	*threads_done = *threads_done + 1;
+	cv_signal(cv);
+	mutex_exit(mtx);
+	zk_thread_exit();
+}
+
+static void
+writer_thread(void *arg)
+{
+
+	int i = 0;
+	int sfd = -1;
+	int count = 0;
+	int nbytes = 0;
+	kmutex_t *mtx;
+	kcondvar_t *cv;
+	int *threads_done;
+	struct data_io *io;
+	worker_args_t *warg = (worker_args_t *)arg;
+
+	sfd = warg->sfd;
+	mtx = warg->mtx;
+	cv = warg->cv;
+	threads_done = warg->threads_done;
+
+	io = kmem_alloc((sizeof (struct data_io) +
+	    warg->io_block_size), KM_SLEEP);
+
+	io->hdr.version = REPLICA_VERSION;
+	io->hdr.opcode = ZVOL_OPCODE_HANDSHAKE;
+	io->hdr.len    = strlen(ds) + 1;
+	strncpy(io->buf, ds, io->hdr.len);
+
+	count = write(sfd, (void *)&(io->hdr), sizeof (zvol_io_hdr_t));
+	if (count == -1) {
+		printf("Sending HDR failed\n");
+		goto exit;
+	}
+
+	count = write(sfd, (void *)(io->buf), io->hdr.len);
+	if (count == -1) {
+		printf("Sending volname is failed\n");
+		goto exit;
+	}
+
+	bzero(io, sizeof (struct data_io));
+	populate(io->buf, warg->io_block_size);
+
+	printf("Start writing ........\n");
+	/* Write data */
+	while (i < warg->max_iops) {
+		io->hdr.version = REPLICA_VERSION;
+		io->hdr.opcode = ZVOL_OPCODE_WRITE;
+		io->hdr.io_seq = i;
+		io->hdr.len    = warg->io_block_size;
```

Comment:
how about random io sizes?

Reply:
Random IO size can be added as an additional Unit test case. Will add one more test case.

Reply:
same testcase also should be fine, as this random IO sizes cover more.. anything should be ok..